File: backend/app/schemas/contracts.py
from typing import Any, Literal
from datetime import datetime
from pydantic import BaseModel, Field
class KPI(BaseModel):
    key:str; label:str; value:float|int|str; unit:str|None=None; delta:float|None=None
class ChartSpec(BaseModel):
    type:Literal["line","bar","area","pie","table","none"]="none"; title:str=""; x_key:str|None=None; y_keys:list[str]=Field(default_factory=list); rows:list[dict[str,Any]]=Field(default_factory=list)
class SourceRef(BaseModel):
    source_id:str; title:str; location:str|None=None; snippet:str|None=None; score:float|None=None
class ToolCall(BaseModel):
    tool_name:str; arguments:dict[str,Any]=Field(default_factory=dict); status:Literal["proposed","approved","executed","failed","cancelled"]="proposed"; error:str|None=None
class ChatRequest(BaseModel):
    conversation_id:str|None=None; query_text:str; attachment_ids:list[str]=Field(default_factory=list); voice_mode:bool=False
class ApprovalRequest(BaseModel):
    decision:Literal["approve","cancel"]
class ConversationSummary(BaseModel):
    id:str; title:str; updated_at:datetime; message_count:int
class ConversationMessage(BaseModel):
    id:str; role:Literal["user","assistant"]; content:str; created_at:datetime
class ConversationDetail(BaseModel):
    id:str; title:str; messages:list[ConversationMessage]
class ChatResponse(BaseModel):
    request_id:str; conversation_id:str; message_id:str; answer:str; intent:str
    kpis:list[KPI]=Field(default_factory=list); chart_spec:ChartSpec=Field(default_factory=ChartSpec)
    insights:list[str]=Field(default_factory=list); recommendations:list[str]=Field(default_factory=list)
    forecast:dict[str,Any]|None=None; sources:list[SourceRef]=Field(default_factory=list); tool_calls:list[ToolCall]=Field(default_factory=list)
    requires_approval:bool=False; suggested_questions:list[str]=Field(default_factory=list); error:dict[str,Any]|None=None
class LoginRequest(BaseModel): email:str; password:str
class AuthUser(BaseModel): id:int; email:str; role:str
class AuthResponse(BaseModel): access_token:str; token_type:str="bearer"; user:AuthUser

File: frontend/src/types/contracts.ts
export interface KPI {
  key: string;
  label: string;
  value: number | string;
  unit: string | null;
  delta: number | null;
}

export interface ChartSpec {
  type: "line" | "bar" | "area" | "pie" | "table" | "none";
  title: string;
  x_key: string | null;
  y_keys: string[];
  rows: Record<string, unknown>[];
}

export interface SourceRef {
  source_id: string;
  title: string;
  location: string | null;
  snippet: string | null;
}

export interface ToolCall {
  tool_name: string;
  arguments: Record<string, unknown>;
  status: "proposed" | "approved" | "executed" | "failed" | "cancelled";
  error?: string | null;
}

export interface ChatRequest {
  conversation_id?: string | null;
  query_text: string;
  attachment_ids: string[];
  voice_mode: boolean;
}

export interface ChatResponse {
  request_id: string;
  conversation_id: string;
  message_id: string;
  answer: string;
  intent: string;
  kpis: KPI[];
  chart_spec: ChartSpec;
  insights: string[];
  recommendations: string[];
  forecast: Record<string, unknown> | null;
  sources: SourceRef[];
  tool_calls: ToolCall[];
  requires_approval: boolean;
  suggested_questions: string[];
  error: Record<string, unknown> | null;
}
